### ISSUE TITLE
Only last active window is restored (Ubuntu)

### ISSUE THREAD

**Issue by BayLee4** [microsoft/vscode#134336]
Does this issue occur when all extensions are disabled?: Yes/No

- VS Code Version: 1.60.2
- OS Version: Ubuntu 21.04 / Gnome

Steps to Reproduce:

1. Open two windows (optional : open folder/files, do some modifications and save them), then click on "Quit 2 Windows" to quit VS Code : 

![Screenshot](https://user-images.githubusercontent.com/63376748/135734365-3eaa9542-6288-41de-a8ae-7bde9c48bf71.png)

2. Open VS Code again : only the last active window is reopened (no matter what value is set in `window.restoreWindows`)

However, if VS Code is closed by another mean (like `pkill code`), on the second launch all windows are correctly restored. 
What is causing this behaviour, and how to fix it ?

**Comment by bpasero** [comment]
As far as I recall, that action on Linux literally closes each window one by one, leaving no choice for us to distinguish from a normal close of the window.

**Comment by BayLee4** [comment]
Is there any workaround ?

**Comment by BayLee4** [comment]
I don't think this is true, other applications do not have any trouble with that. For instance Firefox, Chrome and many other correctly restore all windows even if quitting this way.

**Comment by bpasero** [comment]
Let me investigate, but it will probably require changes from Electron.

**Comment by bpasero** [comment]
Looked into it and compared to Windows and macOS: this actually works fine on macOS because the command you trigger from the menu is "Quit" which is the same when using "File > Quit".

Here is what happens on Linux and Windows:
* the command is not triggering the quit flow as "File > Exit" would do
* instead, each window is being closed one by one
* we go ahead and discard any UI state assuming the user discards a window
* upon restart we only restore the most recent window

In order to change this we would somehow need to know that this is triggered:

![image](https://user-images.githubusercontent.com/900690/135817111-4bc99357-08c5-4753-b639-2960ef503137.png)
![image](https://user-images.githubusercontent.com/900690/135817130-bd3c7c33-7178-4200-b9c1-85567642fcca.png)

@deepak1556 @rzhao271 maybe you are aware of an Electron event that is triggered when using any of the 2 commands on Windows and Linux.

**Comment by trusktr** [comment]
Perhaps what we need is an option to make it so that when closing individual windows, they close immediately without prompting to save, and retain their state. That way, whether by using `File > Exit` or by closing windows one at a time, state will be restored. At least this way there won't be work lost due to unexpected behavior.

Closing windows one by one may lead to only one window being open the next time, but at least we would be able to open a folder, and all the state would be restored.

**Comment by deepak1556** [comment]
@bpasero would [close](https://github.com/electron/electron/blob/main/docs/api/browser-window.md#event-close) event on the window help in this case ?

**Comment by bpasero** [comment]
We listen on `close` and it is the starting point for shutting down a single window (as opposed to all windows where a `quit` event is fired):

https://github.com/microsoft/vscode/blob/b92894493e6f971a3c71912d823e699d3a05d643/src/vs/platform/lifecycle/electron-main/lifecycleMainService.ts#L395

**Comment by deepak1556** [comment]
Ah I see the issue, we need a new event or existing `window-all-closed` event to be emitted for the context menu quit action ? What event do we use on Windows and Linux to determine quit action from `File > Exit` ?

**Comment by bpasero** [comment]
We just trigger our quit function manually then:

https://github.com/microsoft/vscode/blob/b92894493e6f971a3c71912d823e699d3a05d643/src/vs/platform/lifecycle/electron-main/lifecycleMainService.ts#L546

I think the issue here is that the action in Linux closes the windows one by one. If we had a new event to override this behaviour and call our quit method it would probably work for us.

**Comment by vanowm** [comment]
So, how does VSCode know to restore all windows when it was killed from task manager or when computer restarted? Why couldn't the same method be applied to "Quit all" command?
